docs: Manage Pod resources (#1253)
redpanda-data/documentation: docs/manage/kubernetes/manage-resources.mdx

:::info
You cannot decrease the number of CPU cores in a running Redpanda cluster.
:::

To see the available resources on the worker nodes that you have provisioned:

```bash
kubectl describe nodes 
```

## Prerequisites

If you're deploying to production, see the [production best practices](../../../deploy/deployment-option/self-hosted/kubernetes/kubernetes-best-practices) for memory, CPU, and storage volumes.

## Configure CPU resources

By default, Redpanda pins its threads to all cores that you allocate to it. The more cores that are allocated to Redpanda, the more throughput it can support. You can configure the CPU resources allocated to each Redpanda broker by overriding the default values in the Helm chart. For a description of CPU resource units, see the [Kubernetes documentation](https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#meaning-of-cpu).

<Tabs groupId="helm-config" queryString>
<TabItem value="values" label="--values">

```yaml title="cpu-cores.yaml"
resources:
  cpu:
    cores: <number-of-cpu-cores>
```

```bash
helm upgrade --install redpanda redpanda/redpanda -n redpanda --create-namespace \

>   --values cpu-cores.yaml

Reviewer: ```suggestion
  --values cpu-cores.yaml --reuse-values
```

```yaml title="cpu-cores.yaml"
resources:
  cpu:
    cores: 200m
```

:::

### Disable pinnning in shared environments

By default, Redpanda pins its threads to all cores that you allocate to it. 
On a dedicated worker node, CPU pinning allows maximum throughput, but in a development environment, where you may run other Pods on the same worker node as Redpanda, CPU pinning can cause your processes to run slower than usual. In shared environments, you can do the following to tell Redpanda not to pin its threads or memory, and reduce the amount of polling it does to a minimum.

:::note
In production environments, you must run Redpanda brokers on dedicated worker nodes. See [Kubernetes Cluster Requirements](../../../deploy/deployment-option/self-hosted/kubernetes/kubernetes-cluster-requirements).
:::

<Tabs groupId="helm-config" queryString>
<TabItem value="values" label="--values">

```yaml title="cpu-cores-overprovisioned.yaml"
resources:
  cpu:
    cores: <number-of-cpu-cores>
    overprovisioned: true
```

```bash
helm upgrade --install redpanda redpanda/redpanda -n redpanda --create-namespace \

>   --values cpu-cores-overprovisioned.yaml

Reviewer: ```suggestion
  --values cpu-cores-overprovisioned.yaml --reuse-values
```

```

</TabItem>
<TabItem value="flags" label="--set">

```bash
helm upgrade --install redpanda redpanda/redpanda -n redpanda --create-namespace \
  --set resources.cpu.cores=<number-of-cpu-cores> \
  --set resources.cpu.overprovisioned=true
```

</TabItem>
</Tabs>

:::note
<ValuesYaml path="resources.cpu"/>
:::

## Configure memory resources

> On a worker node, you can have the following running processes: Kubernetes, Redpanda, and the Seastar subsystem that Redpanda is built on. Each of these processes consumes memory. You can configure the memory resources that are allocated to these processes.

Reviewer 2: I would be cautious with the statement:
> you can have the following running processes: Kubernetes, Redpanda, and the Seastar subsystem that Redpanda is built on

One might think that Redpanda schedules/runs another process. That's not a case, as Redpanda is single binary.

Author: @RafalKorepta would this be better?

On a worker node, Kubernetes processes and Redpanda processes are running at the same time, including the Seastar subsystem that is built into the Redpanda binary. Each of these processes consumes memory. You can configure the memory resources that are allocated to these processes.

Reviewer 2: Perfect 👍

> On a worker node, you can have the following running processes: Kubernetes, Redpanda, and the Seastar subsystem that Redpanda is built on. Each of these processes consumes memory. You can configure the memory resources that are allocated to these processes.

Author: ```suggestion
On a worker node, Kubernetes processes and Redpanda processes are running at the same time, including the Seastar subsystem that is built into the Redpanda binary. Each of these processes consumes memory. You can configure the memory resources that are allocated to these processes.
```

By default, the Helm chart allocates 80% of the configured memory to Redpanda, leaving the rest for the Seastar subsystem and other container processes. Redpanda recommends this default setting.

:::note
Although you can also allocate the exact amount of memory for Redpanda and the Seastar subsystem manually, Redpanda does not recommend this approach because setting the wrong values can lead to performance issues, instability, or data loss. As a result, this approach is not documented here.
:::

> Redpanda supports single decimal numbers and only the following memory resource units: k M G Ki Mi Gi. For a description of memory resource units, see the [Kubernetes documentation](https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#meaning-of-memory).

Reviewer: What are "single decimal numbers"?

I also don't think this is true for the helm chart. We convert any SI unit at any precision to the nearest megabyte.

Author: this came from the chart: https://github.com/redpanda-data/helm-charts/blob/276201e82d0537e628cc9b5bd5d54613611a48ca/charts/redpanda/values.yaml#L213

Author: We should rephrase to 'supports one decimal place'

Author: @joejulian is the comment in the chart incorrect?

Reviewer: The comment is incorrect. [This function](https://github.com/redpanda-data/helm-charts/blob/8acf816277f3a22d057491117d538ccca24a4bc0/charts/redpanda/templates/_helpers.tpl#L183-L214) will return bytes, regardless of the decimal places of the SI unit.

For memory units, we then convert that number to the nearest whole MiB

<Tabs groupId="helm-config" queryString>
<TabItem value="values" label="--values">

```yaml title="memory.yaml"
resources:
  memory:
    container:
      # Omit min to set it to the same value as max.
      # min:
      max: <number><unit>
```

```bash
helm upgrade --install redpanda redpanda/redpanda -n redpanda --create-namespace \

>   --values memory.yaml

Reviewer: ```suggestion
  --values memory.yaml --reuse-values
```
